kernels/cpu_dense/src/actions/CPUDEvaluateCompartments.cpp

```diff
@@ -30,56 +30,30 @@
  */
 
 
-#include <readdy/kernel/cpu_dense/actions/CPUDCompartments.h>
+#include <readdy/kernel/cpu_dense/actions/CPUDEvaluateCompartments.h>
 
 namespace readdy {
 namespace kernel {
 namespace cpu_dense {
 namespace actions {
 
-CPUDCompartments::CPUDCompartments(const CPUDKernel *const kernel) : kernel(kernel) {}
+CPUDEvaluateCompartments::CPUDEvaluateCompartments(const CPUDKernel *const kernel) : kernel(kernel) {}
 
-void CPUDCompartments::perform() {
+void CPUDEvaluateCompartments::perform() {
     const auto &ctx = kernel->getKernelContext();
-    long long idx = 0;
+    const auto &compartments = ctx.getCompartments();
     for(auto& e : *kernel->getKernelStateModel().getParticleData()) {
         for (auto i = 0; i < compartments.size(); ++i) {
-            if (compartments[i](e.position())) {
-                if (conversions[i].find(e.type) != conversions[i].end()) {
-                    e.type = conversions[i][e.type];
+            if (compartments[i]->isContained(e.position())) {
+                const auto conversions = compartments[i]->getConversions();
+                if (conversions.find(e.type) != conversions.end()) {
+                    e.type = conversions.at(e.type);
```

Comment:
like this you need two map lookups, why not save the iterator of find and dereference it?